state/machine_upgradeseries.go

```diff
@@ -507,32 +507,34 @@ func (m *Machine) GetUpgradeSeriesMessages() ([]string, bool, error) {
 	if err != nil {
 		return nil, false, errors.Trace(err)
 	}
+	// finished means that a subsequent call to this method, while the
+	// Machine Lock is of a similar Machine Status, would return no
+	// additional messages (notifications). Since the value of this variable
+	// is returned, callers may choose to close streams or stop watchers
+	// based on this information.
+	finished := lock.MachineStatus == model.UpgradeSeriesCompleted ||
+		lock.MachineStatus == model.UpgradeSeriesPrepareCompleted
 	// Filter seen messages
 	unseenMessages := make([]UpgradeSeriesMessage, 0)
 	for _, upgradeSeriesMessage := range lock.Messages {
 		if !upgradeSeriesMessage.Seen {
 			unseenMessages = append(unseenMessages, upgradeSeriesMessage)
 		}
 	}
+	if len(unseenMessages) == 0 {
+		return []string{}, finished, nil
+	}
 	sort.Slice(unseenMessages, func(i, j int) bool {
 		return unseenMessages[i].Timestamp.Before(unseenMessages[j].Timestamp)
 	})
 	messages := make([]string, 0)
 	for _, unseenMessage := range unseenMessages {
 		messages = append(messages, unseenMessage.Message)
 	}
-	err = m.SetUpgradeSeriesMessagesAsSeen(unseenMessages)
```

Comment:
Thanks for the original fix, but doesn't this cause us to set things that are already set?
Or is the issue that we were only looking at the in-memory representation and didn't notice that we hadn't recorded that the message was seen?

Reply:
This does set the status of all the messages every time. The reason for this that we need the index of lock message in the array to construct the database transaction. 

Once a user asks to fetch the current messages that are in the db we pull those messages into memory and send them across the wire. We consider those messages to be seen that have been pulled into memory. Therefore we can consider all of the in-memory messages to have been 'seen'.